go/vt/vtctl/grpcvtctldserver/query.go

```diff
@@ -0,0 +1,300 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package grpcvtctldserver
+
+import (
+	"fmt"
+	"time"
+
+	"vitess.io/vitess/go/mysql/collations"
+	"vitess.io/vitess/go/protoutil"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+	"vitess.io/vitess/go/vt/vtctl/schematools"
+
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	vtctldatapb "vitess.io/vitess/go/vt/proto/vtctldata"
+	"vitess.io/vitess/go/vt/proto/vttime"
+)
+
+// 2023-08-05 07:15:16
+const mysqlTimestampLayout = "2006-01-02 15:04:05"
+
+// TODO: explicit column list to guarantee parsing stability
+func selectSchemaMigrationsQuery(condition, order, skipLimit string) string {
+	return fmt.Sprintf(`select
+		*
+		from _vt.schema_migrations where %s %s %s`, condition, order, skipLimit)
+}
+
+func rowToSchemaMigration(row sqltypes.Row) (sm *vtctldatapb.SchemaMigration, err error) {
+	row = row[1:] // We're off by 1 because of the extra Tablet field added to the grouped query result
+
+	sm = new(vtctldatapb.SchemaMigration)
+	sm.Uuid = row[1].ToString()
+	sm.Keyspace = row[2].ToString()
+	sm.Shard = row[3].ToString()
+	sm.Schema = row[4].ToString()
+	sm.Table = row[5].ToString()
+	sm.MigrationStatement = row[6].ToString()
+
+	if name := row[7].ToString(); name != "" {
+		sm.Strategy, err = schematools.ParseSchemaMigrationStrategy(name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	sm.Options = row[8].ToString()
+
+	sm.AddedAt, err = valueToVTTime(row[9])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.RequestedAt, err = valueToVTTime(row[10])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.ReadyAt, err = valueToVTTime(row[11])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.StartedAt, err = valueToVTTime(row[12])
+	if err != nil {
+		return nil, err
+	}
+
+	// liveness_indicator row[13]
```

Comment:
this is still missing as i'm not sure what it is/is for

Reply:
`vitess_liveness_indicator` is a column used internally for keeping track of `vreplication` "liveness", so this is only relevant for `vreplication` migrations, ie `ALTER TABLE` with `online/vitess` strategy.

`vreplication` liveness is based on the last update received by `vreplication`: either a processed binlog entry, or a heartbeat, or a new read transaction. It basically means `vreplication` is ongoing and not stuck.

The column is pretty much internal because it is then used to set `liveness_timestamp`. We can choose to export it or to hide it, at your discretion.

Reply:
Let's include it in the proto and server response, and we can decide later to include it in whether the client actually shows it to the user. But at least it will be in the protobuf definition so folks with more advanced needs can pry into it (and, it sounds like it's not going anywhere so we're not going to need to deprecate/remove from the proto message anytime soon)